### PR TITLE
Remove Container - added "removeVolumes" and "force" options

### PR DESCRIPTION
**Pull request by stepan-romankov**
Stop Container: minor refactoring
Wait Container: added exitCode option

### CODE REVIEW

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerRemoveContainer.groovy [hunk 1]
```diff
@@ -15,10 +15,34 @@
...
 package com.bmuschko.gradle.docker.tasks.container
 
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.Optional
+
 class DockerRemoveContainer extends DockerExistingContainer {
+
+    @Input
+    @Optional
+    boolean removeVolumes
```
Comment (bmuschko) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerRemoveContainer.groovy]: Can you make it a `Boolean` type?

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerRemoveContainer.groovy [hunk 2]
```diff
@@ -15,10 +15,34 @@
...
+
+    @Input
+    @Optional
+    boolean force
```
Comment (bmuschko) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerRemoveContainer.groovy]: Can you make it a `Boolean` type?

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerRemoveContainer.groovy [hunk 3]
```diff
@@ -15,10 +15,34 @@
...
         logger.quiet "Removing container with ID '${getContainerId()}'."
-        dockerClient.removeContainerCmd(getContainerId()).exec()
+        def containerCommand = dockerClient.removeContainerCmd(getContainerId())
+        setContainerCommandConfig(containerCommand)
+        containerCommand.exec()
+    }
+
+    private void setContainerCommandConfig(containerCommand) {
+        if(removeVolumes) {
+            containerCommand.withRemoveVolumes(removeVolumes);
```
Comment (bmuschko) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerRemoveContainer.groovy]: Needs to call getter method to enable it use to use convention mapping.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerRemoveContainer.groovy [hunk 4]
```diff
@@ -15,10 +15,34 @@
...
+        }
+
+        if(force) {
```
Comment (bmuschko) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerRemoveContainer.groovy]: Needs to call getter method to enable it use to use convention mapping.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerCreateContainer.groovy [hunk 5]
```diff
@@ -113,7 +113,7 @@ class DockerCreateContainer extends AbstractDockerRemoteApiTask {
 
     @Input
     @Optional
-    Map<String, Integer> exposedPorts
+    Map<Integer, String> exposedPorts
```
Comment (bmuschko) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerCreateContainer.groovy]: Why was this change made?

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerCreateContainer.groovy [hunk 6]
```diff
@@ -232,7 +232,7 @@ class DockerCreateContainer extends AbstractDockerRemoteApiTask {
         }
 
         if(getExposedPorts()) {
-            def createdExposedPorts = getExposedPorts().collect { threadContextClassLoader.createExposedPort(it.key, it.value) }
+            def createdExposedPorts = getExposedPorts().collect { threadContextClassLoader.createExposedPort(it.value, it.key) }
```
Comment (bmuschko) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerCreateContainer.groovy]: Why was this changed?

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerWaitContainer.groovy [hunk 7]
```diff
@@ -16,9 +16,17 @@
...
+
+    DockerWaitContainer() {
+        ext.getExitCode = { exitCode }
+    }
+
     @Override
     void runRemoteCommand(dockerClient) {
         logger.quiet "Waiting for container with ID '${getContainerId()}'."
-        dockerClient.waitContainerCmd(getContainerId()).exec()
+        exitCode = dockerClient.waitContainerCmd(getContainerId()).exec()
```
Comment (bmuschko) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerWaitContainer.groovy]: I think it's fair to assume that we want to fail the build if the exit code != 0. I don't think we necessarily need to expose the exit code. What's the use case?

**Comment by bmuschko** [comment]
For all of this I'd also like to see a functional test.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerCreateContainer.groovy (hunk 5 above)
Comment (stepan-romankov) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerCreateContainer.groovy]: There is very problematic to define in configuration which ports should be exposed cause it need Multimap (in case when you expose multiple TCP ports). And Groovy/Gradle  doesn't have multi map implementation in class path.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerWaitContainer.groovy (hunk 7 above)
Comment (stepan-romankov) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerWaitContainer.groovy]: It's very useful if you want to execute unit tests inside container. exitCode is easy way to determine if tests are passed or not.

**Comment by stepan-romankov** [comment]
Hello, I've made functional tests. Please apply pull request.

**Comment by stepan-romankov** [comment]
up

**Comment by bmuschko** [comment]
Can we break this up into multiple pull requests? I do not agree with the change to `exposedPorts`. We'll need to solve this in a different way. We can break it up by task type:

- DockerCreateContainer
- DockerRemoveContainer
- DockerWaitContainer

Can you close this pull request and create three new ones? That'll make the merge much easier.

**Comment by bmuschko** [comment]
Some of these things have been provided by different pull requests. I am not going to merge this pull request as the requested changes haven't been made.
